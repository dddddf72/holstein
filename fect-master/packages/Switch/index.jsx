import { computed, ref, toRefs, watch } from 'vue'
import { validator, theme, createNameSpace } from '../utils'

const [createComponent] = createNameSpace('Switch')
const { normalSizes } = theme

import './switch.less'

export default createComponent({
  props: {
    checked: Boolean,
    size: {
      type: String,
      validator: validator.enums(normalSizes),
      default: 'medium',
    },
    disabled: Boolean,
  },
  emits: ['change'],
  setup(props, { attrs, slots, emit }) {
    const { checked, size, disabled } = toRefs(props)
    const selfChecked = ref(checked.value)
    const changeHandler = (e) => {
      const selfEvent = {
        target: {
          checked: !selfChecked.value,
        },
        stopPropagation: e.stopPropagation,
        preventDefault: e.preventDefault,
        nativeEvent: e,
      }
      selfChecked.value = !selfChecked.value
      emit('change', selfEvent)
    }
    const calcStatus = computed(() => {
      let str = ''
      selfChecked.value && (str += ' checked')
      disabled.value && (str += ' disabled')
      return str
    })

    return () => (
      <>
        <label
          {...attrs}
          className={`fect-switch ${size.value} ${calcStatus.value}
          ${attrs.class ? attrs.class : ''}`}
        >
          <input
            className={`fect-switch-checkBox ${size.value}`}
            type="checkBox"
            checked={selfChecked.value}
            disabled={disabled.value}
            onChange={changeHandler}
          ></input>
          <div
            className={`fect-swtich-slider ${size.value} ${calcStatus.value}`}
          ></div>
        </label>
      </>
    )
  },
})
