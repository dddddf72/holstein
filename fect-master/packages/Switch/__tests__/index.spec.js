import { mount } from '@vue/test-utils'
import Switch from '../index'

describe('Switch', () => {
  it('should be render as element', () => {
    const swtich = mount(Switch)
    expect(swtich.html()).toMatchSnapshot()
    expect(() => swtich.unmount()).not.toThrow()
  })

  it('should be support different sizes', () => {
    const wrapper = mount({
      setup() {
        return { sizesList: ['mini', 'small', 'medium', 'large'] }
      },
      render(h) {
        return (
          <>
            {this.sizesList.map((item) => (
              <Switch size={item} key={item} />
            ))}
          </>
        )
      },
    })
    const els = wrapper.findAll('.fect-switch')
    expect(wrapper.html()).toMatchSnapshot()
    expect(els[0].classes('mini')).toBe(true)
    expect(els[1].classes('small')).toBe(true)
    expect(els[2].classes('medium')).toBe(true)
    expect(els[3].classes('large')).toBe(true)
  })

  it('should be support disbaled', () => {
    const wrapper = mount(<Switch disabled={true} />)
    const el = wrapper.find('.fect-swtich-slider')
    expect(el.classes('disabled')).toBe(true)
  })

  it('should emit event change', async () => {
    const wrapper = mount(<Switch />)
    const el = wrapper.find('.fect-swtich-slider')
    expect(el.classes('checked')).toBe(false)
    await wrapper.find('[type="checkBox"]').trigger('change')
    expect(el.classes('checked')).toBe(true)
  })
})
