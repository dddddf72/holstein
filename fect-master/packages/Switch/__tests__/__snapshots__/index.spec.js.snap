// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should be render as element 1`] = `
<label class="fect-switch medium 
          "><input class="fect-switch-checkBox medium" type="checkBox">
  <div class="fect-swtich-slider medium "></div>
</label>
`;

exports[`Switch should be support different sizes 1`] = `
<label class="fect-switch mini 
          "><input class="fect-switch-checkBox mini" type="checkBox">
  <div class="fect-swtich-slider mini "></div>
</label><label class="fect-switch small 
          "><input class="fect-switch-checkBox small" type="checkBox">
  <div class="fect-swtich-slider small "></div>
</label><label class="fect-switch medium 
          "><input class="fect-switch-checkBox medium" type="checkBox">
  <div class="fect-swtich-slider medium "></div>
</label><label class="fect-switch large 
          "><input class="fect-switch-checkBox large" type="checkBox">
  <div class="fect-swtich-slider large "></div>
</label>
`;
